File: Java/Hangman/Hangman.java
import java.util.Random;
import java.util.ArrayList;
import java.util.Scanner;
import java.util.Set;
import java.util.HashSet;

public class Hangman {

    public static void main(String[] args) {
        // generate the random word from a list of random words
        String[] words = {"generous", "curiosity", "continental", "captivate", "ubiquitous"};

        Random random = new Random();
        int randomIndex = random.nextInt(words.length);

        String randomWord = words[randomIndex];
        // System.out.println(randomWord);

        // create the underscores list
        ArrayList<String> underscores = new ArrayList<String>();
        for (int i = 0; i < randomWord.length(); i++) {
            underscores.add("_");
        }
        String result = String.join(" ", underscores);
        System.out.println(result);

        // while loop for game
        int guesses = 7;

        // create scanner for getting input
        Scanner sc = new Scanner(System.in);

        // create Set to keep track of letters guessed
        Set<Character> lettersGuessed = new HashSet<>();

        while (guesses > 0) {
            System.out.println("Number of guesses: " + guesses);
            System.out.print("Enter a letter: ");
            char letter = sc.next().charAt(0);

            while (lettersGuessed.contains(letter)) {
                System.out.println(String.format("You already guessed %c. Guess a different letter.", letter));
                System.out.print("Enter a letter: ");
                letter = sc.next().charAt(0);
            }

            // add the new letter to the lettersGuessed Set
            lettersGuessed.add(letter);

            // System.out.println("Random word length: " + randomWord.length());
            boolean guessed = false;
            for (int i = 0; i < randomWord.length(); i++) {
                // System.out.println("Comparing " + letter + " to " + randomWord.charAt(i));
                if (letter == randomWord.charAt(i)) {
                    underscores.set(i, Character.toString(letter));
                    guessed = true;
                }
            }
            result = String.join(" ", underscores);
            System.out.println(result);

            if (guessed == false) {
                    guesses--;
            }

            // contains is "in" Python equivalent
            if (!underscores.contains("_")) {
                System.out.println("You win!");
                break;
            }
            else if (guesses == 0) {
                System.out.println("Sorry you lose. The word was " + randomWord);
            }

        }
        
        sc.close();
    }
    
}

File: Python/Hangman/hangman.py
import random

words = ["anything", "science", "english", "kitten", "burger", "fanta", "propensity", "swell", "calculation", "dress to impress"]

random_word = random.choice(words)
# print(random_word)

def create_hangman_list(word):
    hangman_list = []
    for letter in word:
        if letter == ' ':
            hangman_list.append(' ')
        else:
            hangman_list.append('_')   
    return hangman_list

def create_phrase(hangman_list):
    return ' '.join(hangman_list)

def update_hangman_list(hangman_list, indices, guess):
    for i in indices:
        hangman_list[i] = guess
        
def check_for_win(hangman_list):
    # if '_' not in hangman_list:
    #     return True
    # return False
    
    if '_' in hangman_list:
        return False
    else:
        return True

attempts = 7
hangman_list = create_hangman_list(random_word)
hangman_phrase = create_phrase(hangman_list)
letters_guessed = []

while attempts > 0:
    print(f"\nNumber of attempts left: {attempts}")
    print(hangman_phrase)
    guess = input("Guess a letter: ")
    
    if guess in letters_guessed:
        print("You already guessed this letter. Please choose another letter.")
        print(f"Letters already guessed: {', '.join(letters_guessed)}")
        # print("Letters already guessed:", ', '.join(letters_guessed))
        continue
    else:
        letters_guessed.append(guess)
    
    if guess in random_word:
        indices = []
        for i in range(len(random_word)): 
            if random_word[i] == guess: 
               indices.append(i)
        update_hangman_list(hangman_list, indices, guess)
        hangman_phrase = create_phrase(hangman_list)
    else:
        attempts -= 1
    
    # win condition    
    if check_for_win(hangman_list):
        print(f"\nYou win with {attempts} attempts left!")
        # print("You win with", attempts, "attempts left!")
        print(f"The phrase was {random_word}")
        # print("The phrase was", random_word)
        break
    
if not check_for_win(hangman_list):
    print("\nYou lost :(")
    print(f"The phrase was {random_word}")